File: src/domain/models/persona.py
"""Domain models for the Tool Translation bounded context -- persona management.

PersonaType enumerates the agent persona archetypes.
Register classifies personas as technical or non-technical.
PersonaDefinition is a frozen value object describing one agent persona.
PERSONA_REGISTRY maps every PersonaType to its canonical PersonaDefinition.
SUPPORTED_TOOLS and TOOL_TARGET_PATHS define valid tool identifiers and
their target file path templates.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass

from src.domain.models.errors import InvariantViolationError


class PersonaType(enum.Enum):
    """Agent persona archetypes supported by alty."""

    SOLO_DEVELOPER = "solo_developer"
    TEAM_LEAD = "team_lead"
    AI_TOOL_SWITCHER = "ai_tool_switcher"
    PRODUCT_OWNER = "product_owner"
    DOMAIN_EXPERT = "domain_expert"


class Register(enum.Enum):
    """Communication register for a persona -- technical or non-technical."""

    TECHNICAL = "technical"
    NON_TECHNICAL = "non_technical"


@dataclass(frozen=True)
class PersonaDefinition:
    """Immutable value object describing one agent persona.

    Attributes:
        name: Human-readable persona name.
        persona_type: The archetype this persona represents.
        register: Communication register (technical / non-technical).
        description: Short description of the persona's role.
        instructions_template: Template text for the persona's agent instructions.
    """

    name: str
    persona_type: PersonaType
    register: Register
    description: str
    instructions_template: str

    def __post_init__(self) -> None:
        if not self.name or not self.name.strip():
            raise InvariantViolationError("Persona name cannot be empty")
        if not self.instructions_template or not self.instructions_template.strip():
            raise InvariantViolationError("Persona instructions template cannot be empty")


PERSONA_REGISTRY: dict[PersonaType, PersonaDefinition] = {
    PersonaType.SOLO_DEVELOPER: PersonaDefinition(
        name="Solo Developer",
        persona_type=PersonaType.SOLO_DEVELOPER,
        register=Register.TECHNICAL,
        description="Full-stack developer following DDD + TDD + SOLID principles",
        instructions_template=(
            "# Solo Developer Agent\n\n"
            "You are a solo developer agent responsible for implementing features\n"
            "using DDD, TDD, and SOLID principles. Follow Red/Green/Refactor strictly.\n"
            "Write failing tests first, then minimal code to pass, then refactor."
        ),
    ),
    PersonaType.TEAM_LEAD: PersonaDefinition(
        name="Team Lead",
        persona_type=PersonaType.TEAM_LEAD,
        register=Register.TECHNICAL,
        description="Technical lead responsible for architecture and code quality",
        instructions_template=(
            "# Team Lead Agent\n\n"
            "You are a team lead agent responsible for architecture review\n"
            "and DDD compliance. Ensure bounded context boundaries are respected\n"
            "and code quality meets project standards."
        ),
    ),
    PersonaType.AI_TOOL_SWITCHER: PersonaDefinition(
        name="AI Tool Switcher",
        persona_type=PersonaType.AI_TOOL_SWITCHER,
        register=Register.TECHNICAL,
        description="Agent specialized in cross-tool configuration and migration",
        instructions_template=(
            "# AI Tool Switcher Agent\n\n"
            "You are an AI tool switcher agent responsible for cross-tool\n"
            "configuration mapping and config translation. Ensure consistent\n"
            "behavior across supported AI coding tools."
        ),
    ),
    PersonaType.PRODUCT_OWNER: PersonaDefinition(
        name="Product Owner",
        persona_type=PersonaType.PRODUCT_OWNER,
        register=Register.NON_TECHNICAL,
        description="Business-focused agent for requirements and prioritization",
        instructions_template=(
            "# Product Owner Agent\n\n"
            "You are a product owner agent focused on business requirements\n"
            "and prioritization. Use business language and ensure requirements\n"
            "are captured clearly for the development team."
        ),
    ),
    PersonaType.DOMAIN_EXPERT: PersonaDefinition(
        name="Domain Expert",
        persona_type=PersonaType.DOMAIN_EXPERT,
        register=Register.NON_TECHNICAL,
        description="Domain knowledge specialist for business rules and terminology",
        instructions_template=(
            "# Domain Expert Agent\n\n"
            "You are a domain expert agent responsible for business rules\n"
            "and domain terminology. Ensure ubiquitous language is used correctly\n"
            "and domain invariants are properly captured."
        ),
    ),
}


SUPPORTED_TOOLS: tuple[str, ...] = ("claude-code", "cursor", "roo-code", "opencode")

TOOL_TARGET_PATHS: dict[str, str] = {
    "claude-code": ".claude/agents/{name}.md",
    "cursor": ".cursor/rules/{name}.mdc",
    "roo-code": ".roo-code/modes/{name}.md",
    "opencode": ".opencode/agents/{name}.md",
}

File: internal/shared/domain/valueobjects/persona.go
package valueobjects

import (
	"fmt"
	"strings"

	domainerrors "github.com/alty-cli/alty/internal/shared/domain/errors"
)

// PersonaType enumerates agent persona archetypes.
type PersonaType string

// Persona type constants.
const (
	PersonaSoloDeveloper  PersonaType = "solo_developer"
	PersonaTeamLead       PersonaType = "team_lead"
	PersonaAIToolSwitcher PersonaType = "ai_tool_switcher"
	PersonaProductOwner   PersonaType = "product_owner"
	PersonaDomainExpert   PersonaType = "domain_expert"
)

// AllPersonaTypes returns all valid persona type values.
func AllPersonaTypes() []PersonaType {
	return []PersonaType{
		PersonaSoloDeveloper,
		PersonaTeamLead,
		PersonaAIToolSwitcher,
		PersonaProductOwner,
		PersonaDomainExpert,
	}
}

// Register classifies personas as technical or non-technical.
type Register string

// Register constants.
const (
	RegisterTechnical    Register = "technical"
	RegisterNonTechnical Register = "non_technical"
)

// PersonaDefinition is an immutable value object describing one agent persona.
type PersonaDefinition struct {
	name                 string
	personaType          PersonaType
	register             Register
	description          string
	instructionsTemplate string
}

// NewPersonaDefinition creates a validated PersonaDefinition.
func NewPersonaDefinition(
	name string,
	personaType PersonaType,
	register Register,
	description string,
	instructionsTemplate string,
) (*PersonaDefinition, error) {
	if strings.TrimSpace(name) == "" {
		return nil, fmt.Errorf("persona name cannot be empty: %w", domainerrors.ErrInvariantViolation)
	}
	if strings.TrimSpace(instructionsTemplate) == "" {
		return nil, fmt.Errorf("persona instructions template cannot be empty: %w", domainerrors.ErrInvariantViolation)
	}
	return &PersonaDefinition{
		name:                 name,
		personaType:          personaType,
		register:             register,
		description:          description,
		instructionsTemplate: instructionsTemplate,
	}, nil
}

// Name returns the persona name.
func (p *PersonaDefinition) Name() string { return p.name }

// PersonaType returns the persona archetype.
func (p *PersonaDefinition) PersonaType() PersonaType { return p.personaType }

// Register returns the communication register.
func (p *PersonaDefinition) Register() Register { return p.register }

// Description returns the persona description.
func (p *PersonaDefinition) Description() string { return p.description }

// InstructionsTemplate returns the agent instructions template.
func (p *PersonaDefinition) InstructionsTemplate() string { return p.instructionsTemplate }

// PersonaRegistry returns the canonical registry of all persona definitions.
func PersonaRegistry() map[PersonaType]*PersonaDefinition {
	solo, _ := NewPersonaDefinition(
		"Solo Developer", PersonaSoloDeveloper, RegisterTechnical,
		"Full-stack developer following DDD + TDD + SOLID principles",
		"# Solo Developer Agent\n\nYou are a solo developer agent responsible for implementing features\nusing DDD, TDD, and SOLID principles. Follow Red/Green/Refactor strictly.\nWrite failing tests first, then minimal code to pass, then refactor.",
	)
	lead, _ := NewPersonaDefinition(
		"Team Lead", PersonaTeamLead, RegisterTechnical,
		"Technical lead responsible for architecture and code quality",
		"# Team Lead Agent\n\nYou are a team lead agent responsible for architecture review\nand DDD compliance. Ensure bounded context boundaries are respected\nand code quality meets project standards.",
	)
	switcher, _ := NewPersonaDefinition(
		"AI Tool Switcher", PersonaAIToolSwitcher, RegisterTechnical,
		"Agent specialized in cross-tool configuration and migration",
		"# AI Tool Switcher Agent\n\nYou are an AI tool switcher agent responsible for cross-tool\nconfiguration mapping and config translation. Ensure consistent\nbehavior across supported AI coding tools.",
	)
	po, _ := NewPersonaDefinition(
		"Product Owner", PersonaProductOwner, RegisterNonTechnical,
		"Business-focused agent for requirements and prioritization",
		"# Product Owner Agent\n\nYou are a product owner agent focused on business requirements\nand prioritization. Use business language and ensure requirements\nare captured clearly for the development team.",
	)
	de, _ := NewPersonaDefinition(
		"Domain Expert", PersonaDomainExpert, RegisterNonTechnical,
		"Domain knowledge specialist for business rules and terminology",
		"# Domain Expert Agent\n\nYou are a domain expert agent responsible for business rules\nand domain terminology. Ensure ubiquitous language is used correctly\nand domain invariants are properly captured.",
	)

	return map[PersonaType]*PersonaDefinition{
		PersonaSoloDeveloper:  solo,
		PersonaTeamLead:       lead,
		PersonaAIToolSwitcher: switcher,
		PersonaProductOwner:   po,
		PersonaDomainExpert:   de,
	}
}

// SupportedTools returns the tuple of supported tool identifiers.
func SupportedTools() []string {
	return []string{"claude-code", "cursor", "roo-code", "opencode"}
}

// ToolTargetPaths returns the mapping of tool identifiers to file path templates.
func ToolTargetPaths() map[string]string {
	return map[string]string{
		"claude-code": ".claude/agents/{name}.md",
		"cursor":      ".cursor/rules/{name}.mdc",
		"roo-code":    ".roo-code/modes/{name}.md",
		"opencode":    ".opencode/agents/{name}.md",
	}
}
